specs/agents/tracing-instrumentation-azure-functions.md

```diff
@@ -0,0 +1,103 @@
+# Tracing Azure Functions
+
+An Azure Function Application can implement one or more handler functions that are executed via different trigger mechanisms (e.g. HTTP).
+Depending on these trigger mechanisms and the actual [Azure Functions runtime language and version](https://learn.microsoft.com/en-us/azure/azure-functions/supported-languages),
+the signature of these handler functions will vary, of course.
+
+In general however, there is always a *generic context* object available (e.g. [`FunctionContext`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.azure.functions.worker.functioncontext)) that provides metadata and context for an instrumentation.
+
+Depending on the actual kind of invocation (trigger type) there might be an additional object to retrieve metadata and context from.
+In case of an HTTP invocation/trigger there is e.g. [`HttpRequestData`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.identitymodel.protocols.httprequestdata).
+
+## Generic Instrumentation
+
+In general, to instrument an Azure Functions application, we create transactions that wrap the execution of its handler methods. In cases where we cannot assume any trigger type or extract trigger-specific data (e.g. if the trigger type is unsupported),
+we wrap the handler method with a transaction, while using the generic context object to derive necessary fields:
+
+| Field | Value | Description | Source |
+| - | - | - | - |
+| `name` | e.g. `MySampleTrigger` | The transaction name. Use function name if trigger type is `other`. | *generic context* |
+| `type` | e.g. `request`, `messaging` | The transaction type. | Use `request` if trigger type is undefined. |
+| `outcome` | `success`/`failure` | Set to `failure` if a function error can be detected otherwise `success`. | |
+| `result` | `success`/`failure` |Set to `failure` if a function error can be detected, otherwise `success`. | Trigger specific. |
+| `faas.name` | e.g. `MySampleTrigger` | The function name. | *generic context* |
+| `faas.id` | e.g. `/subscriptions/d2ba53be-0815-4...` | The [fully qualified resource ID](https://learn.microsoft.com/en-us/rest/api/resources/resources/get-by-id) of the Azure Function, which has this format: `/subscriptions/<SUBSCRIPTION_GUID>/resourceGroups/<RG>/providers/Microsoft.Web/sites/<FUNCAPP>/functions/<FUNC>` | *generic context*, environment |
+| `faas.trigger.type` | `other` | The trigger type. Use `other` if the trigger type is unknown or cannot be specified. | More concrete triggers are `http`, `pubsub`, `datasource`, `timer` (see specific triggers below). |
+
+| `faas.execution` | `203621a2-62f...` | The unique invocation id of the function. | *generic context* |
+| `faas.coldstart` | `true`/`false` | A boolean value indicating whether this function invocation was a cold start or not. | [see section below](deriving-cold-starts)
```

Comment:
Yes, see https://github.com/elastic/apm/pull/712

Reply:
I may be wrong, but #712 is about `metadata.cloud.provider` and this is about `transaction.context.cloud.origin.provider`. Two separate (and redundant?) fields.

Reply:
Right, I should have given more detail in my answer above.
Yes, it's a different field but I left it out on purpose, because:
* It's redundant.
* The way it was explained to me, all the `transaction.context.cloud.origin.*` fields represent the invocation side and cover a specific Lambda use case.

Reply:
Makes sense. Thanks for the details.